configuration/vpn.md

```diff
@@ -60,73 +60,128 @@ Using a ProxyVM to set up a VPN client gives you the ability to:
-    This allows your system to connect to the vpn server with the protocol `$PROT` under the port 1194.
-    
-    `iptables -I OUTPUT -o lo -j ACCEPT`  
-    This allows connections from the system to the system.
-
-    `iptables -I FORWARD -o eth0 -j DROP`  
-    `iptables -I FORWARD -i eth0 -j DROP`  
-    This blocks forwarding of connections through your plain network device (in case the vpn tunnel breaks).
-
-    `iptables -t nat -I PR-QBS -p udp --dport 53 -j DNAT --to-destination $DNS`  
-    `iptables -t nat -I PR-QBS -p tcp --dport 53 -j DNAT --to-destination $DNS`  
-    This will rewrite the DNS destination, and the traffic will be routed down the vpn tunnel. (to prevent DNS leaks)
-
+    NOTE: If the connection breaks down all traffic will by default be routed through the upstream network device eth0 (we will stop this with iptables in step 3).
+
+    Also add the following to accomodate a DNS script:
+    ```
+    script-security 2
+    up 'qubes-vpn-handler.sh up'
+    down 'qubes-vpn-handler.sh down'
+    ```
+
+3.  Setup iptables.
+    Edit the firewall script with `sudo nano /rw/config/qubes-firewall-user-script` and add:
+
+	```
+	#!/bin/bash
+	#    First, block all outgoing traffic
+	iptables -P OUTPUT DROP
+    iptables -F OUTPUT
+
+	#    Add the `qvpn` group to system, if it doesn't already exist
+    if ! grep -q "^qvpn:" /etc/group ; then
+        groupadd -rf qvpn
+        sync
+    fi
+    sleep 2s
+
+	#    Allow traffic from the `qvpn` group to the uplink interface (eth0);
+	#    Our openvpn will run as group `qvpn`.
+    iptables -A OUTPUT -p all -o eth0 -m owner --gid-owner qvpn \
+    -m state --state NEW,ESTABLISHED -j ACCEPT
+
+	#	Allow queries to DNS server:
+    iptables -A OUTPUT -p udp -o eth0 --dport 53 -m state --state NEW,ESTABLISHED -j ACCEPT
+    iptables -A OUTPUT -p tcp -o eth0 --dport 53 -m state --state NEW,ESTABLISHED -j ACCEPT
+
+	#	Allow internal system connections:
+    iptables -I OUTPUT -o lo -j ACCEPT
+
+	#	Block forwarding of connections through upstream network device
+	#	(in case the vpn tunnel breaks):
+    iptables -I FORWARD -o eth0 -j DROP  
+    iptables -I FORWARD -i eth0 -j DROP
+    ```
     Now save `/rw/config/qubes-firewall-user-script` and make it executable:  
     `sudo chmod +x /rw/config/qubes-firewall-user-script`
-    
-4.  Setup the vpn's autostart:  
-   Edit to `/rw/config/rc.local`, make it executable  (`sudo chmod +x /rw/config/rc.local`) and add:  
 
-        ln -s /home/user/vpn.cfg /etc/openvpn/vpn.conf;  
-        systemctl --no-block start openvpn@vpn.service;
+4.  Create the DNS-handling script.
+    Use `sudo nano /rw/config/openvpn/qubes-vpn-handler.sh` to edit and add:
+    ```
+    #!/bin/bash
+    set -e
+
+# Pop-up notification variables
+SPID=$(pgrep -U user -f dconf-service)
+dbus=$(grep -z DBUS_SESSION_BUS_ADDRESS /proc/$SPID/environ|cut -d= -f2-)
+export DBUS_SESSION_BUS_ADDRESS=$dbus
```

Comment:
You can get user environment by sourcing `/tmp/qubes-session-env`. Or add `-` to `su` invocation below - to load login scripts, which will source that file.